packages/editor/src/components/post-publish-panel/media-util.js

```diff
@@ -0,0 +1,120 @@
+export function generateUniqueFilenames( urls ) {
+	const basenames = new Set();
+
+	return urls.map( ( url ) => {
+		// We prefer to match the remote filename, if possible.
+		const filename = getFilename( url );
+		let basename = '';
+		let extension = '';
+
+		if ( filename ) {
+			const parts = filename.split( '.' );
+			if ( parts.length > 1 ) {
+				// Assume the last part is the extension.
+				extension = parts.pop();
+			}
+			basename = parts.join( '.' );
+		}
+
+		if ( ! basename ) {
+			// It looks like we don't have a basename, so let's use a UUID.
+			basename = uuid();
+		}
+
+		if ( basenames.has( basename ) ) {
+			// Append a UUID to deduplicate the basename.
+			// The server will try to deduplicate on its own if we don't do this,
+			// but it may run into a race condition
+			// (see https://github.com/WordPress/gutenberg/issues/64899).
+			// Deduplicating the filenames before uploading is safer.
+			basename = `${ basename }-${ uuid() }`;
+		}
+
+		basenames.add( basename );
+
+		return { url, basename, extension };
+	} );
+}
+
+const mimeTypeToExtension = {
+	'image/apng': 'apng',
+	'image/avif': 'avif',
+	'image/bmp': 'bmp',
+	'image/gif': 'gif',
+	'image/vnd.microsoft.icon': 'ico',
+	'image/jpeg': 'jpg',
+	'image/jxl': 'jxl',
+	'image/png': 'png',
+	'image/svg+xml': 'svg',
+	'image/tiff': 'tiff',
+	'image/webp': 'webp',
+	'video/x-msvideo': 'avi',
+	'video/mp4': 'mp4',
+	'video/mpeg': 'mpeg',
+	'video/ogg': 'ogg',
+	'video/mp2t': 'ts',
+	'video/webm': 'webm',
+	'video/3gpp': '3gp',
+	'video/3gpp2': '3g2',
+};
+
+// Get the file extension to use for a given mime type.
+export function getExtensionFromMimeType( mime ) {
+	mime = ( mime ?? '' ).toLowerCase();
+
+	let extension = mimeTypeToExtension[ mime ];
+
+	if ( ! extension ) {
+		// We don't know which extension to use, so we need to fall back to
+		// something safe. The server should replace it with an appropriate
+		// extension for the mime type.
+		if ( mime.startsWith( 'image/' ) ) {
+			extension = 'png';
+		}
+		if ( mime.startsWith( 'video/' ) ) {
+			extension = 'mp4';
+		}
+	}
+
+	// If all else fails, try an empty extension.
+	// The server will probably reject the upload, but there isn't much
+	// else we can do.
+	return extension || '';
+}
+
+// Returns an array of { url, blobPromise } objects, where the promise
+// points to a fetched blob. Blobs will have unique filenames.
+export function fetchMedia( urls ) {
+	return generateUniqueFilenames( urls ).map(
+		( { url, basename, extension } ) => {
+			const blobPromise = window
+				.fetch( url.includes( '?' ) ? url : url + '?' )
+				.then( ( response ) => response.blob() )
+				.then( ( blob ) => {
+					// Not all remote filenames have an extension, but we need to
+					// provide one, or the server is likely to reject the upload.
+					if ( ! extension ) {
+						extension = getExtensionFromMimeType( blob.type );
+					}
+					blob.name = `${ basename }.${ extension }`;
+					return blob;
```

Comment:
[`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob) doesn't have a `name` property. What you want is a [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) object.

```suggestion
					return blob;
					return new File( [ blob ], `${ basename }.${ extension }`, { type: blob.type } )
```

Reply:
Yes, you're absolutely right 👍 It's best to do the right thing and instantiate and actual `File` object instead of using duck-typing here; who knows what the browser might do behind the scenes.

I took the opportunity to rename the property as well, for clarity.

Done in b93954dba144aa69aa941e134b0d3d89a5eec56c.